packages/libs/error-stack/src/ser.rs

```diff
@@ -0,0 +1,152 @@
+//! Implementation of general report serialization.
+//!
+//! The value can be of any type, currently only printable attachments and context are supported, in
+//! the near future any values will be supported through the use of hooks.
+//!
+//! The returned JSON returned is a list of all current sources with the following output:
+//!
+//! ```json
+//! {
+//!     "context": "string value",
+//!     "attachments": ["all attachments leading up to this context"],
+//!     "sources": [] // recursive render using `frame.sources()`
+//! }
+//! ```
+
+use alloc::{format, vec, vec::Vec};
+
+use serde::{ser::SerializeMap, Serialize, Serializer};
+
+use crate::{AttachmentKind, Context, Frame, FrameKind, Report};
+
+struct SerializeAttachment<'a>(&'a Frame);
+
+impl<'a> Serialize for SerializeAttachment<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let Self(frame) = self;
+
+        #[allow(clippy::match_same_arms)]
+        match frame.kind() {
+            FrameKind::Context(_) => {
+                // `SerializeContext` ensures that no context is ever serialized
+                unreachable!()
```

Comment:
Do we want to store `AttachmentKind` instead of `Frame` then?

Reply:
This is in prep for hooks, where we will end up `request_ref` and `downcast_ref` (like with hooks).

Reply:
clarified in [0414ac3](https://github.com/hashintel/hash/pull/1290/commits/0414ac33a2e77fd9a0b6d8675329e9d52573cab4)